README.md
# Jekyll Avatar

*A Jekyll plugin for rendering GitHub avatars*

[![CI](https://github.com/benbalter/jekyll-avatar/actions/workflows/ci.yml/badge.svg)](https://github.com/benbalter/jekyll-avatar/actions/workflows/ci.yml)

Jekyll Avatar makes it easy to add GitHub avatars to your Jekyll site by specifying a username. If performance is a concern, Jekyll Avatar is deeply integrated with the GitHub avatar API, ensuring avatars are cached and load in parallel. It even automatically upgrades users to Retina images, when supported.

## Installation

Add the following to your site's `Gemfile`:

```ruby
gem 'jekyll-avatar'
```

And add the following to your site's `_config.yml` file:

```yaml
plugins:
  - jekyll-avatar
```
Note: If you are using a Jekyll version less than 3.5.0, use the `gems` key instead of `plugins`.

## Usage

Simply add the following, anywhere you'd like a user's avatar to appear:

```
{% avatar [USERNAME] %}
```

With `[USERNAME]` being the user's GitHub username:

```
{% avatar hubot %}
```

That will output:

```html
<img class="avatar avatar-small" src="https://avatars3.githubusercontent.com/hubot?v=3&amp;s=40" alt="hubot" srcset="https://avatars3.githubusercontent.com/hubot?v=3&amp;s=40 1x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=80 2x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=120 3x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=160 4x" width="40" height="40" />
```

### Customizing

You can customize the size of the resulting avatar by passing the size argument:

```
{% avatar hubot size=50 %}
```

That will output:

```html
<img class="avatar" src="https://avatars3.githubusercontent.com/hubot?v=3&amp;s=50" alt="hubot" srcset="https://avatars3.githubusercontent.com/hubot?v=3&amp;s=50 1x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=100 2x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=150 3x, https://avatars3.githubusercontent.com/hubot?v=3&amp;s=200 4x" width="50" height="50" />
```

### Passing the username as variable

You can also pass the username as a variable, like this:

```
{% assign user="hubot" %}
{% avatar {{ user }} %}
```

Or, if the variable is someplace a bit more complex, like a loop:

```
{% assign employees = "alice|bob" | split:"|" %}
{% for employee in employees %}
  {% avatar user=employee %}
{% endfor %}
```

### Lazy loading images

For pages showing a large number of avatars, you may want to load the images lazily.

```liquid
{% avatar hubot lazy=true %}
```

This will set the `data-src` and `data-srcset` attributes on the `<img>` tag, which is compatible with many lazy load JavaScript plugins, such as:

* https://www.andreaverlicchi.eu/lazyload/
* https://appelsiini.net/projects/lazyload/

### Using with GitHub Enterprise

To use Jekyll Avatars with GitHub Enterprise, you must set the `PAGES_AVATARS_URL` environmental variable.

This should be the full URL to the avatars subdomain or subpath. For example:

* With subdomain isolation: `PAGES_AVATARS_URL="https://avatars.github.example.com"`
* Without subdomain isolation: `PAGES_AVATARS_URL="https://github.example.com/avatars"`


History.markdown
## 0.8.0 / 2022-04-15

### Minor Enhancements

  * Use `Kernel#format` to render `<img />` HTML tag (#46)
  * Check if username and size matches a pattern once (#48)

### Bug Fixes

  * Reduce allocations from computing username (#44)
  * Stringify keys of `:attributes` hash (#42)
  * Parse tag markup once per instance (#40)
  * Compute `:srcset` with an array of integer strings (#43)
  * Assign string values for attributes (#47)
  * Parse only custom-host provided through ENV (#45)

### Development Fixes

  * Profile memory usage from rendering avatars (#41)
  * Bundle only relevant files in the gem (#50)
  * Upgrade to GitHub-native Dependabot (#52)
  * Remove redundant specifications (#56)
  * Improve context in workflow job names (#57)
  * Remove `@benbalter`-specific community health files (#58)
  * Update gem specification (#60)
  * Add workflow to release gem via GH Actions (#63)

### Documentation

  * Fix typo in README.md (#62)

## 0.7.0 / 2019-08-12

Refer [`v0.7.0` Release Note](https://github.com/jekyll/jekyll-avatar/releases/tag/v0.7.0) to see
what changed since previous release.


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)

require "jekyll"
require "jekyll-avatar"
require "rspec-html-matchers"

TEST_DIR = File.dirname(__FILE__)
TMP_DIR  = File.expand_path("../tmp", TEST_DIR)

RSpec.configure do |config|
  config.include RSpecHtmlMatchers
end

def doc_with_content(_content, opts = {})
  my_site = site(opts)
  options = { :site => my_site, :collection => collection(my_site) }
  doc = Jekyll::Document.new(source_dir("_test/doc.md"), options)
  doc.merge_data!({ "author" => "hubot2" })
  doc
end

def tmp_dir(*files)
  File.join(TMP_DIR, *files)
end

def source_dir(*files)
  tmp_dir("source", *files)
end

def dest_dir(*files)
  tmp_dir("dest", *files)
end

def collection(site, label = "test")
  Jekyll::Collection.new(site, label)
end

def site(opts = {})
  defaults = Jekyll::Configuration::DEFAULTS
  opts = opts.merge(
    "source"      => source_dir,
    "destination" => dest_dir
  )
  conf = Jekyll::Utils.deep_merge_hashes(defaults, opts)
  Jekyll::Site.new(conf)
end

def fixture(name)
  path = File.expand_path "./fixtures/#{name}.json", File.dirname(__FILE__)
  File.open(path).read
end

def with_env(key, value)
  old_value = ENV[key]
  ENV[key] = value
  yield
ensure
  ENV[key] = old_value
end


spec/jekyll/avatar_spec.rb
# frozen_string_literal: true

require "spec_helper"

describe Jekyll::Avatar do
  subject { described_class.parse "avatar", text, tokenizer, parse_context }

  let(:doc) { doc_with_content(content) }
  let(:username) { "hubot" }
  let(:args) { "" }
  let(:text) { "#{username} #{args}".squeeze(" ") }
  let(:content) { "{% avatar #{text} %}" }
  let(:rendered) { subject.render(nil) }
  let(:tokenizer) { Liquid::Tokenizer.new("") }
  let(:parse_context) { Liquid::ParseContext.new }
  let(:output) do
    doc.content = content
    doc.output  = Jekyll::Renderer.new(doc.site, doc).run
  end

  let(:server_number) { 3 }
  let(:host) { "https://avatars#{server_number}.githubusercontent.com" }
  let(:uri) do
    uri = Addressable::URI.parse(host)
    uri.path = "#{uri.path}/#{username}"
    uri.query_values = { :v => 3, :s => width }.to_a
    uri
  end
  let(:height) { "40" }
  let(:width) { "40" }
  let(:scales) { (1..4) }
  let(:src) { src_hash["1x"] }
  let(:src_hash) do
    scales.map do |scale|
      uri.query_values = { "v" => 3, "s" => width.to_i * scale }.to_a
      ["#{scale}x", uri.to_s]
    end.to_h
  end
  let(:srcset) { src_hash.map { |scale, src| "#{src} #{scale}" }.join(", ") }

  it "has a version number" do
    expect(Jekyll::Avatar::VERSION).not_to be nil
  end

  it "outputs the HTML" do
    expect(output).to have_tag("p") do
      with_tag "img", :with => {
        "alt"                 => username,
        "class"               => "avatar avatar-small",
        "data-proofer-ignore" => "true",
        "height"              => height,
        "src"                 => src,
        "srcset"              => srcset,
        "width"               => width,
      }
    end
  end

  it "parses the username" do
    expect(subject.send(:username)).to eql(username)
  end

  it "determines the server number" do
    expect(subject.send(:server_number)).to eql(server_number)
  end

  it "builds the host" do
    expect(subject.send(:host)).to eql(host)
  end

  context "with a custom host" do
    let(:host) { "http://avatars.example.com" }

    context "with subdomain isolation" do
      it "builds the host" do
        with_env("PAGES_AVATARS_URL", host) do
          expect(subject.send(:host)).to eql(host)
        end
      end

      it "builds the URL" do
        with_env("PAGES_AVATARS_URL", host) do
          expect(subject.send(:url)).to eql(src)
        end
      end
    end

    context "without subdomain isolation" do
      let(:host) { "http://github.example.com/avatars" }

      it "builds the URL" do
        with_env("PAGES_AVATARS_URL", host) do
          expect(subject.send(:url)).to eql(src)
        end
      end
    end
  end

  it "builds the path" do
    expect(subject.send(:path)).to eql("hubot?v=3&s=40")
  end

  it "defaults to the default size" do
    expect(subject.send(:size)).to eql(width)
  end

  it "builds the URL" do
    expect(subject.send(:url)).to eql(src)
  end

  it "builds the params" do
    attrs = subject.send(:attributes)
    expect(attrs).to eql({
      :"data-proofer-ignore" => "true",
      :class                 => "avatar avatar-small",
      :alt                   => username,
      :src                   => src,
      :srcset                => srcset,
      :width                 => width,
      :height                => height,
    })
  end

  it "includes data-proofer-ignore" do
    expect(output).to have_tag "img", :with => {
      "data-proofer-ignore" => "true",
    }
  end

  context "retina" do
    it "builds the path with a scale" do
      expect(subject.send(:path, 2)).to eql("hubot?v=3&s=80")
    end

    context "when given a scale" do
      let(:width) { "80" }

      it "builds the URL with a scale" do
        expect(subject.send(:url, 2)).to eql(src)
      end
    end

    it "builds the srcset" do
      srcset = subject.send(:srcset)
      src_hash.each do |scale, url|
        regex = Regexp.escape("#{url} #{scale}")
        expect(srcset).to match(regex)
      end
    end
  end

  context "when passed @hubot as a username" do
    let(:username) { "@hubot" }

    it "parses the username" do
      expect(subject.send(:username)).to eql("hubot")
    end
  end

  context "with a size is passed" do
    let(:width) { "45" }
    let(:args) { "size=#{width}" }

    it "parses the user's requested size" do
      expect(subject.send(:size)).to eql(width)
    end
  end

  context "with a size < 48" do
    it "includes the avatar-small class" do
      expect(rendered).to match(%r!avatar-small!)
    end

    it "calculates the classes" do
      expect(subject.send(:classes)).to eql("avatar avatar-small")
    end
  end

  context "with a size > 48" do
    let(:width) { "80" }
    let(:args) { "size=#{width}" }

    it "doesn't include the avatar-small class" do
      expect(rendered).not_to match(%r!avatar-small!)
    end

    it "calculates the classes" do
      expect(subject.send(:classes)).to eql("avatar")
    end
  end

  context "when passed the username as a rendered variable" do
    let(:username) { "hubot2" }
    let(:server_number) { 0 }
    let(:content) { "{% assign user='#{username}' %}{% avatar {{ user }} %}" }

    it "parses the variable" do
      expect(output).to have_tag "img", :with => { "src" => src }
    end
  end

  context "when passed the username as a variable-argument" do
    let(:username) { "hubot2" }
    let(:server_number) { 0 }
    let(:content) { "{% assign user='hubot2' %}{% avatar user=user %}" }

    it "parses the variable" do
      expect(output).to have_tag "img", :with => { "src" => src }
    end
  end

  context "when passed the username as a sub-variable-argument" do
    let(:username) { "hubot2" }
    let(:server_number) { 0 }
    let(:content) { "{% avatar user=page.author %}" }

    it "parses the variable" do
      expect(output).to have_tag "img", :with => { "src" => src }
    end
  end

  context "loops" do
    let(:content) do
      <<~LIQUID
        {% assign users = "a|b" | split:"|" %}
        {% for user in users %}
          {% avatar user=user %}
        {% endfor %}
      LIQUID
    end

    it "renders each avatar" do
      expect(output).to have_tag("p") do
        with_tag "img", :with => { "alt" => "a" }
        with_tag "img", :with => { "alt" => "b" }
      end
    end
  end

  context "lazy loading" do
    let(:args) { "lazy=true" }

    it "sets the image URL as the data-src" do
      expect(output).to have_tag "img", :with => {
        "src"         => "",
        "data-src"    => src,
        "data-srcset" => srcset,
      }, :without => {
        "srcset" => %r!.*!,
      }
    end
  end
end


lib/jekyll/avatar.rb
require 'jekyll/avatar/version'

module Jekyll
  class Avatar < Liquid::Tag
    SERVERS = 4
    DEFAULT_SIZE = 40
    VERSION = 3

    def initialize(tag_name, text, tokens)
      super
      @text = text
    end

    def render(_context)
      tag = '<img '

      # See http://primercss.io/avatars/#small-avatars
      if size < 48
        tag << 'class="avatar avatar-small" '
      else
        tag << 'class="avatar" '
      end

      tag << "src=\"#{url}\" alt=\"#{username}\" "
      tag << "width=\"#{size}\" height=\"#{size}\" "
      tag << '/>'
      tag
    end

    private

    def username
      @username ||= @text.split(' ').first.sub('@', '')
    end

    def size
      @size ||= begin
        matches = @text.match(/\bsize=(\d+)\b/i)
        matches ? matches[1].to_i : DEFAULT_SIZE
      end
    end

    def path
      @path ||= "#{username}?v=#{VERSION}&s=#{size}"
    end

    def server_number
      @server_number ||= Zlib.crc32(path) % SERVERS
    end

    def host
      @host ||= "avatars#{server_number}.githubusercontent.com"
    end

    def url
      @url ||= "https://#{host}/#{path}"
    end
  end
end

Liquid::Template.register_tag('avatar', Jekyll::Avatar)


lib/jekyll-avatar.rb
# frozen_string_literal: true

require "zlib"

module Jekyll
  class Avatar < Liquid::Tag
    def self.generate_template_with(keys)
      attrs = (BASE_ATTRIBUTES + keys).map! { |key| %(#{key}="%<#{key}>s") }.join(" ")
      "<img #{attrs} />"
    end
    private_class_method :generate_template_with

    #

    SERVERS      = 4
    DEFAULT_SIZE = "40"
    API_VERSION  = "3"

    BASE_ATTRIBUTES = %w(
      class alt width height data-proofer-ignore src
    ).freeze

    BASE_TEMPLATE = generate_template_with %w(srcset)
    LAZY_TEMPLATE = generate_template_with %w(data-src data-srcset)

    private_constant :BASE_ATTRIBUTES, :BASE_TEMPLATE, :LAZY_TEMPLATE

    def initialize(_tag_name, text, _tokens)
      super
      @text = text.strip
      @markup = Liquid::Template.parse(@text)

      @size = compute_size
      @user_variable = extract_user_variable

      @custom_host = ENV["PAGES_AVATARS_URL"]
      @custom_host = "" unless @custom_host.is_a?(String)
    end

    def render(context)
      @context = context
      @text    = @markup.render(@context)
      template = lazy_load? ? LAZY_TEMPLATE : BASE_TEMPLATE
      format(template, attributes)
    end

    private

    def attributes
      result = {
        :class                 => classes,
        :alt                   => username,
        :width                 => size,
        :height                => size,
        :"data-proofer-ignore" => "true"
      }

      if lazy_load?
        result[:src] = ""
        result[:"data-src"] = url
        result[:"data-srcset"] = srcset
      else
        result[:src] = url
        result[:srcset] = srcset
      end

      result
    end

    def lazy_load?
      @text.include?("lazy=true")
    end

    def extract_user_variable
      matches = @text.match(%r!\buser=([\w\.]+)\b!)
      matches[1] if matches
    end

    def username
      return lookup_variable(@user_variable) if @user_variable

      result = @text.include?(" ") ? @text.split(" ")[0] : @text
      result.start_with?("@") ? result.sub("@", "") : result
    end

    # Lookup a Liquid variable in the current context.
    #
    # variable - the variable name, as a string.
    #
    # Returns the value of the variable in the context or the variable name if not found.
    def lookup_variable(variable)
      lookup = @context
      if variable.include?(".")
        variable.split(".").each do |value|
          lookup = lookup[value]
        end
      else
        lookup = lookup[variable]
      end

      lookup || variable
    end

    # Returns a string value
    def compute_size
      matches = @text.match(%r!\bsize=(\d+)\b!i)
      matches ? matches[1] : DEFAULT_SIZE
    end
    attr_reader :size

    def path(scale = 1)
      "#{username}?v=#{API_VERSION}&s=#{scale == 1 ? size : (size.to_i * scale)}"
    end

    def server_number
      Zlib.crc32(path) % SERVERS
    end

    def host
      if @custom_host.empty?
        "https://avatars#{server_number}.githubusercontent.com"
      else
        @custom_host
      end
    end

    def parsed_host
      @parsed_host ||= {}
      @parsed_host[host] ||= Addressable::URI.parse(host)
    end

    def url_with_custom_host(scale = 1)
      uri = parsed_host
      uri.path << "/" unless uri.path.end_with?("/")
      uri = uri.join path(scale)
      uri.to_s
    end

    def url(scale = 1)
      return url_with_custom_host(scale) unless @custom_host.empty?

      "#{host}/#{path(scale)}"
    end

    SCALES = %w(1 2 3 4).freeze
    private_constant :SCALES

    def srcset
      SCALES.map { |scale| "#{url(scale.to_i)} #{scale}x" }.join(", ")
    end

    # See http://primercss.io/avatars/#small-avatars
    def classes
      size.to_i < 48 ? "avatar avatar-small" : "avatar"
    end
  end
end

Liquid::Template.register_tag("avatar", Jekyll::Avatar)
